docs/design/2019-04-11-indexmerge.md

```diff
@@ -0,0 +1,320 @@
+		return nil
+	}
+	return CreateIndexMergeIntersectionPath(partialPaths, tableFilters)
+}
+
+// Now, we just use all path in partialPaths to generate a IndexMergeIntersection.
+// We also need to merge possible paths. 
+// For example: 
+//   index:  ix1(a)
+//   condition： a > 1 and a < 10
+//   we will get two partial paths and they all use index ix1. 
+// IndexMergePath.tableFilters: 
+//   <1> tableFilters
+//   <2> after merge operation, if any indexPath's tableFilter is not nil, we 
+//       should add indexPath’s tableFilter into IndexMergePath.tableFilters
+CreateIndexMergeIntersectionPath(partialPaths, tableFilters) {
+}
+
+```
+	
+### Executor
+Graph bellow illustrates execution of IndexMerge scan. 
+<img alt="Execution Model" src="./imgs/execution_model.png" width="500pt"/>
+
+Every index plan in `PhysicalIndexMergeLookUpReader` will start an `IndexWorker` to execute the IndexScan plan and send handles to AndOrWorker. AndOrWorker is responsible for doing set operations (and, or) to get final handles. Then `AndOrWoker` sends final handles to `TableWokers` to get rows from TiKV.
+
+We can take some tricks to make execution at pipeline mode without considering the order.
+
+
+(1)	IndexMergeIntersection
+  
+ <img alt="IndexMergeIntersection" src="./imgs/and.jpeg" width="500pt"/>
+   We take an example to explain it. Use set1 to record rowids which are returned by ix1 but not sent to tableWorker. Use set2 to record the same thing for ix2.
+If new rowid comes from ix1, first we check if it is in set2. If so, we delete it from set2 and send it to tableWorker. Otherwise, we add it into set1. For the above figure, we use the following table to show the processing.
+
+| new rowid | set1 | set2 | sent to TableWorker |
+| :------:| :------: | :------: | :------: |
+| 2(ix1) | [2] | [ ] | [ ] |
+| 1(ix2) | [2] | [1] | [ ] |
+| 5(ix1) | [2,5] |[1] | [ ] |
+| 5(ix2) | [2] | [1] | [5] |
+| 7(ix1) | [2,7] |[1] | [ ] |
+| 2(ix2) | [7] | [1] | [2] |
+
+
+
+(2)	IndexMergeUnion
+
+
+   We take a structure(we call it set) to record which rowids are accessed. If a new rowid returned by IndexScan, check if it is in set. If in it, we just skip it. Otherwise, we add it into set and send it to tableWorker. 
+
+### Cost Model
+Cost model will consider three factors: IO, CPU, and Network. 
+
+- `IndexMergeType` = 1
+
+	IO Cost = (totalRowCount + mergedRowCount) * scanFactor
+	
+	Network Cost = (totalRowCount + mergedRowCount) * networkFactor
+	
+	Cpu Memory Cost = totalRowCount * cpuFactor + totalRowCount * memoryFactor
+	
+- `IndexMergeType` = 2
+
+	IO Cost = (totalRowCount) * scanFactor
+	
+	Network Cost = totalRowCount * networkFactor
+	
+	Cpu Memory Cost = totalRowCount * cpuFactor + totalRowCount * memoryFactor
+	
+- `IndexMergeType` = 3
+
+	IO Cost = (totalRowCount + mergedRowCount) * scanFactor
+	
+	Network Cost = (totalRowCount + mergedRowCount) * networkFactor
+	
+	Cpu Memory Cost = totalRowCount * cpuFactor + mergedRowCount * memoryFactor

```

Comment:
we should also add the cost for `set` access

Reply:
ok. Whether add cost for insert and delete?

Reply:
We should consider cost for set access, while the detailed formula can be figured out later during implementation?

Reply:
then update this document?

Reply:
@eurekaka can we use Cpu Memory Cost to represent the cost for set operation?

Reply:
@hailanwhu yes we should.

Reply:
I mean the `Cpu Memory Cost` in proposal now.

Reply:
@zz-jason I use `Cpu Memory Cost` in the proposal to represent the cost of set operations.

Reply:
Got it. Seems we need to refine the cost calculation for the present IndexReader/IndexLookupReader to let it possible to chose this IndexMergeUnion operator.

Reply:
yes. Or may use new factor for Set operations. However it is better to refine the cost calculation.